### PR TITLE
chore: remove circular dependency

### PR DESCRIPTION
**Pull request by FrankEssenberger**
Closes SAP/cloud-sdk-backlog#38

There was a circular dependency between the planned packages:
- http-client provided the `executeHttpRequest` method
- this method could fetch destinations
- the fetch destination method used the `executeHttpRequest` again.

In order to overcome this the `executeHttpRequest` was replaced with a plain axios request in the `xsuaa-service` and `destination-service`. I also created a new folder (to be package) called `http-agent` this contains all parts of the http agents and proxy generation and will be used by the `http-client` and services as well.

### CODE REVIEW

File: packages/core/src/http-client/index.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 export * from './http-client';
 export * from './http-client-types';
-export * from './http-agent';
-export * from './agent-config';
+export * from '../http-agent/http-agent';
+export * from '../http-agent/agent-config';
```
Comment (jjtang1985) [inline packages/core/src/http-client/index.ts:4]:
```suggestion
```

File: packages/core/src/http-agent/http-agent.ts [hunk 2]
```diff
@@ -172,3 +178,23 @@ export function getUrlProtocol(destination: Destination): Protocol | undefined {
...
+export function urlAndAgent(
+  targetUri: string
+): Pick<HttpRequestConfig, 'baseURL' | 'httpAgent' | 'httpsAgent' | 'proxy'> {
+  let destination: Destination = { url: targetUri, proxyType: 'Internet' };
+  if (proxyStrategy(destination) === ProxyStrategy.INTERNET_PROXY) {
+    destination = addProxyConfigurationInternet(destination);
+  }
+  return {
+    baseURL: destination.url,
+    proxy: false,
```
Comment (jjtang1985) [inline packages/core/src/http-agent/http-agent.ts:197]: question: why setting hard coded `false` to proxy?

File: packages/core/src/connectivity/scp-cf/xsuaa-service.ts [hunk 3]
```diff
@@ -268,28 +253,40 @@ const tokenKeyKeyMapping: { [key: string]: keyof TokenKey } = {
...
       ? tokenServiceUrlOrXsuaaServiceCredentials
       : getTokenServiceUrl(tokenServiceUrlOrXsuaaServiceCredentials);
 
-  let destination: Destination = { url: targetUri, proxyType: 'Internet' };
-  if (proxyStrategy(destination) === ProxyStrategy.INTERNET_PROXY) {
-    destination = addProxyConfigurationInternet(destination);
-  }
+  const config: AxiosRequestConfig = {
+    ...urlAndAgent(targetUri),
+    proxy: false,
```
Comment (jjtang1985) [inline packages/core/src/connectivity/scp-cf/xsuaa-service.ts:280]: in line 279, the proxy value is set to `false`, so we can remove duplicate value.
I have another question below about the value `false`.

File: packages/core/src/connectivity/scp-cf/xsuaa-service.ts [hunk 4]
```diff
@@ -268,28 +253,40 @@ const tokenKeyKeyMapping: { [key: string]: keyof TokenKey } = {
...
+
 function post(
   tokenServiceUrlOrXsuaaServiceCredentials: string | XsuaaServiceCredentials,
   authHeader: string,
   body: string,
   options: ResilienceOptions = { enableCircuitBreaker: true }
-): Promise<HttpResponse> {
-  const config = wrapXsuaaPostRequestHeader(authHeader, body);
+): Promise<AxiosResponse> {
+  // const config = wrapXsuaaPostRequestHeader(authHeader, body);
```
Comment (jjtang1985) [inline packages/core/src/connectivity/scp-cf/xsuaa-service.ts:272]:
```suggestion
```

File: packages/core/src/connectivity/scp-cf/xsuaa-service.spec.ts [hunk 5]
```diff
@@ -39,19 +38,27 @@ describe('xsuaa', () => {
...
-        proxyType: 'Internet',
-        proxyConfiguration: {
+      const expectedConfig: AxiosRequestConfig = {
+        baseURL: 'https://provider.example.com/oauth/token',
+        method: 'post',
+        data: 'grant_type=client_credentials',
+        proxy: false,
+        headers: {
+          Accept: 'application/json',
+          Authorization: 'Basic aG9yc3RpOmJvcnN0aQ==',
```
Comment (jjtang1985) [inline packages/core/src/connectivity/scp-cf/xsuaa-service.spec.ts:50]: minor: in general, I would reuse the variable defined above.

File: packages/core/src/connectivity/scp-cf/destination/destination-service.ts [hunk 6]
```diff
@@ -232,32 +219,41 @@ function errorMessageFromResponse(error: AxiosError): string {
     : '';
 }
 
+function getAxiosConfig(uri, headers): AxiosRequestConfig {
```
Comment (jjtang1985) [inline packages/core/src/connectivity/scp-cf/destination/destination-service.ts:222]: not used?

**Review by jjtang1985 — APPROVED** [review]
LGTM. I left some minor comments.

File: packages/core/src/connectivity/scp-cf/xsuaa-service.spec.ts (hunk 5 above)
Comment (FrankEssenberger) [reply packages/core/src/connectivity/scp-cf/xsuaa-service.spec.ts:50]: yes good point

**Pull request merged**